lib/auth/trustedcluster.go

```diff
@@ -112,6 +120,76 @@ func (a *AuthServer) DeleteTrustedCluster(name string) error {
 	return nil
 }
 
+// addEnabledTrustedCluster does Trusted Cluster exchange and creates enabled
+// resources on the backend.
+func (a *AuthServer) addEnabledTrustedCluster(t services.TrustedCluster) error {
+	// do the trust exchange. if establishTrust is successful, the remote
+	// auth server has verified out token.
+	remoteCAs, err := a.establishTrust(t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// add remote ca to our backend
+	err = a.addRemoteCAs(remoteCAs, t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// add reverse tunnel to our backend
+	reverseTunnel := services.NewReverseTunnel(
+		t.GetName(),
+		[]string{t.GetReverseTunnelAddress()},
+	)
+	err = a.UpsertReverseTunnel(reverseTunnel)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// update trusted cluster resource
+	err = a.Presence.UpsertTrustedCluster(t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// addDisabledTrustedCluster does Trusted Cluster exchange and creates disabled
+// resources on the backend.
+func (a *AuthServer) addDisabledTrustedCluster(t services.TrustedCluster) error {
+	// do the trust exchange. if establishTrust is successful, the remote
+	// auth server has verified out token.
+	remoteCAs, err := a.establishTrust(t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// add remote ca to our backend
+	err = a.addRemoteCAs(remoteCAs, t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// deactivate the ca we just added
+	err = a.DeactivateCertAuthority(services.CertAuthID{Type: services.UserCA, DomainName: t.GetName()})
```

Comment:
if you fail here you will have cert authority created but no trusted cluster associated with it, why not create trusted cluster resource first?